Libs/MRML/Core/vtkMRMLModelNode.cxx

```diff
@@ -141,64 +173,130 @@ void vtkMRMLModelNode::SetAndObservePolyData(vtkPolyData *polyData)
       }
 
     vtkNew<vtkTrivialProducer> tp;
-    tp->SetOutput(polyData);
+    tp->SetOutput(mesh);
     // Propagate ModifiedEvent onto the trivial producer to make sure
-    // PolyDataModifiedEvent is triggered.
+    // MeshModifiedEvent is triggered.
     if (!this->DataEventForwarder)
       {
       this->DataEventForwarder = vtkEventForwarderCommand::New();
       }
     this->DataEventForwarder->SetTarget(tp.GetPointer());
-    polyData->AddObserver(vtkCommand::ModifiedEvent, this->DataEventForwarder);
-    this->SetPolyDataConnection(tp->GetOutputPort());
+    mesh->AddObserver(vtkCommand::ModifiedEvent, this->DataEventForwarder);
+
+    if (polydata)
+      {
+      this->SetPolyDataConnection(tp->GetOutputPort());
+      }
+    else
+      {
+      this->SetUnstructuredGridConnection(tp->GetOutputPort());
+      }
     }
 }
 
+//----------------------------------------------------------------------------
+void vtkMRMLModelNode::SetAndObservePolyData(vtkPolyData *polyData)
+{
+  this->SetAndObserveMesh(polyData);
+}
+
 //---------------------------------------------------------------------------
-vtkPolyData* vtkMRMLModelNode::GetPolyData()
+vtkPointSet *vtkMRMLModelNode::GetMesh()
 {
-  vtkAlgorithm* producer = this->PolyDataConnection ?
-    this->PolyDataConnection->GetProducer() : 0;
-  return vtkPolyData::SafeDownCast(
+  vtkAlgorithm* producer = this->MeshConnection ?
+    this->MeshConnection->GetProducer() : 0;
+  return vtkPointSet::SafeDownCast(
     producer ? producer->GetOutputDataObject(
-      this->PolyDataConnection->GetIndex()) : 0);
+      this->MeshConnection->GetIndex()) : 0);
+}
+
+//---------------------------------------------------------------------------
+vtkPolyData* vtkMRMLModelNode::GetPolyData()
+{
+  if (vtkPolyData* polydata = vtkPolyData::SafeDownCast(this->GetMesh()))
+    {
+    return polydata;
+    }
+  else
+    {
+    this->GeometryFilter->Update();
+    return this->GeometryFilter->GetOutput();
+    }
+}
+
+//---------------------------------------------------------------------------
+vtkUnstructuredGrid* vtkMRMLModelNode::GetUnstructuredGrid()
+{
+  return vtkUnstructuredGrid::SafeDownCast(this->GetMesh());
 }
 
 //---------------------------------------------------------------------------
 void vtkMRMLModelNode
-::SetPolyDataConnection(vtkAlgorithmOutput *newPolyDataConnection)
+::SetMeshConnection(vtkAlgorithmOutput *newMeshConnection)
 {
-  if (newPolyDataConnection == this->PolyDataConnection)
+  if (newMeshConnection == this->MeshConnection)
     {
     return;
     }
 
-  vtkAlgorithm* oldPolyDataAlgorithm = this->PolyDataConnection ?
-    this->PolyDataConnection->GetProducer() : 0;
+  vtkAlgorithm* oldMeshAlgorithm = this->MeshConnection ?
+    this->MeshConnection->GetProducer() : 0;
 
-  this->PolyDataConnection = newPolyDataConnection;
+  this->MeshConnection = newMeshConnection;
```

Comment:
vtkSmartPointer only manages register/unregister. vtkSetAndObserveMRMLObjectEventsMacro manages register/unregister and observe/unobserved events as well. So, the best would be to use vtkSetAndObserveMRMLObjectEventsMacro.